テスト test_models.py 181 行目が確率的に落ちる

```
======================================================================
FAIL: test_iterate_classifications (budgetmapper.tests.test_models.BudgetTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/mikan/develop/spendingjp/openspending-backend/backend/budgetmapper/tests/test_models.py", line 181, in test_iterate_classifications
    self.assertEqual(a, e)
AssertionError: {'cla[51 chars]ect (cQCo4EYfiwGjxgDzRs5eqk)>, <Classification[131 chars]wQ)>} != {'cla[51 chars]ect (aVbzh5m5cKesM4H7MuyxR5)>, <Classification[131 chars]FP)>}
Diff is 1051 characters long. Set self.maxDiff to None to see it.

----------------------------------------------------------------------
```

--- a/backend/budgetmapper/models.py
+++ b/backend/budgetmapper/models.py
@@ -99,6 +99,24 @@ class LevelNameListField(ArrayField):
         )
 
 
+class ItemOrderField(models.IntegerField):
+    def __init__(self, *args, **kwargs):
+        super(ItemOrderField, self).__init__(*args, **dict(kwargs, null=True, db_index=True))
+
+    def pre_save(self, model_instance, add):
+        val = getattr(model_instance, self.attname)
+        if val is None:
+            qs = Classification.objects.filter(classification_system=model_instance.classification_system).order_by(
+                "-item_order"
+            )
+            if qs.count() == 0:
+                val = 0
+            else:
+                val = qs[0].item_order + 1
+            setattr(model_instance, self.attname, val)
+        return val
+
+
 class Government(models.Model):
     id = PkField()
     name = NameField()
@@ -119,7 +137,7 @@ class ClassificationSystem(models.Model):
 
     @property
     def roots(self) -> models.QuerySet:
-        return Classification.objects.filter(classification_system=self, parent=None).order_by("code")
+        return Classification.objects.filter(classification_system=self, parent=None).order_by("item_order")
 
     @property
     def leaves(self) -> models.QuerySet:
@@ -147,6 +165,7 @@ class Classification(models.Model):
     code = models.CharField(max_length=64, null=True, db_index=True)
     classification_system = models.ForeignKey(ClassificationSystem, on_delete=models.CASCADE)
     parent = models.ForeignKey("self", on_delete=models.CASCADE, null=True)
+    item_order = ItemOrderField()
     created_at = CurrentDateTimeField()
     updated_at = AutoUpdateCurrentDateTimeField()
 
@@ -165,7 +184,10 @@ class Classification(models.Model):
 
     @property
     def direct_children(self) -> models.QuerySet:
-        return Classification.objects.filter(parent=self).order_by("code")
+        return Classification.objects.filter(parent=self).order_by("item_order")
+
+    class Meta:
+        unique_together = ("classification_system", "item_order")
 
 
 class Budget(models.Model):

--- a/backend/budgetmapper/tests/factories.py
+++ b/backend/budgetmapper/tests/factories.py
@@ -21,6 +21,12 @@ class ClassificationSystemFactory(DjangoModelFactory):
     level_names = ["款", "項", "目", "事業", "節", "節細"]
 
 
+def get_max_item_order_plus_one():
+    q = models.Classification.objects.order_by("-item_order")
+    if q.count() == 0:
+        return 0
+    return q[0].item_order + 1
+
 class ClassificationFactory(DjangoModelFactory):
     class Meta:
         model = models.Classification
@@ -28,6 +34,7 @@ class ClassificationFactory(DjangoModelFactory):
     name = fuzzy.FuzzyText(length=5, suffix="費")
     code = factory.LazyFunction(lambda: None if random.randint(0, 2) % 2 else str(random.randint(1, 15)))
     classification_system = factory.SubFactory(ClassificationSystemFactory)
+    item_order = fuzzy.FuzzyAttribute(get_max_item_order_plus_one)
     parent = None
 
 

--- a/backend/budgetmapper/tests/test_models.py
+++ b/backend/budgetmapper/tests/test_models.py
@@ -4,7 +4,8 @@ from unittest.mock import MagicMock, patch
 
 from budgetmapper import models
 from django.core.exceptions import ValidationError
-from django.test import TestCase
+from django.db.utils import IntegrityError
+from django.test import TestCase, TransactionTestCase
 
 from . import factories
 
@@ -75,7 +76,7 @@ class ClassificationSystemTest(TestCase):
             self.assertEqual(a, e)
 
 
-class ClassificationTest(TestCase):
+class ClassificationTest(TransactionTestCase):
     def test_classification_has_name(self) -> None:
         cs = factories.ClassificationSystemFactory()
         sut = models.Classification(name="総務費", classification_system=cs)
@@ -106,6 +107,7 @@ class ClassificationTest(TestCase):
         sut = models.Classification(
             name="総務雑費", classification_system=root.classification_system, code="3", parent=root
         )
+        sut.save()
         sut.full_clean()
 
         cs1 = factories.ClassificationSystemFactory()
@@ -113,7 +115,30 @@ class ClassificationTest(TestCase):
         root = factories.ClassificationFactory(classification_system=cs1)
         sut = models.Classification(name="不整合あり", classification_system=cs2, parent=root)
         with self.assertRaises(ValidationError):
+            sut.save()
             sut.full_clean()
+
+    def test_classification_has_item_order(self) -> None:
+        cs = factories.ClassificationSystemFactory()
+        sut0 = models.Classification(name="総務費", classification_system=cs)
+        sut1 = models.Classification(name="総務雑貨", classification_system=cs)
+        sut0.save()
+        sut1.save()
+        self.assertEqual(sut0.item_order, 0)
+        self.assertEqual(sut1.item_order, 1)
+        sut2 = models.Classification(name="上書き不可", classification_system=cs, item_order=1)
+        with self.assertRaises(IntegrityError):
+            sut2.save()
+        sut10 = models.Classification(name="議会費", classification_system=cs, item_order=10)
+        sut11 = models.Classification(name="民生貨", classification_system=cs)
+        sut10.save()
+        sut11.save()
+        self.assertEqual(sut10.item_order, 10)
+        self.assertEqual(sut11.item_order, 11)
+        cs2 = factories.ClassificationSystemFactory()
+        sut = models.Classification(name="農林水産業費", classification_system=cs2)
+        sut.save()
+        self.assertEqual(sut.item_order, 0)
 
 
 class BudgetTest(TestCase):

--- a/backend/budgetmapper/tests/test_request.py
+++ b/backend/budgetmapper/tests/test_request.py
@@ -106,6 +106,9 @@ class WdmmgTestCase(BudgetMapperTestUserAPITestCase):
         cl1 = factories.ClassificationFactory(classification_system=cs, code="2")
         cl10 = factories.ClassificationFactory(classification_system=cs, parent=cl1, code="2.1")
         cl100 = factories.ClassificationFactory(classification_system=cs, parent=cl10, code="2.1.1")
+        cl2 = factories.ClassificationFactory(classification_system=cs, code="10")
+        cl20 = factories.ClassificationFactory(classification_system=cs, parent=cl2, code="10.1")
+        cl200 = factories.ClassificationFactory(classification_system=cs, parent=cl20, code="10.1.1")
         bud = factories.BudgetFactory(
             name="まほろ市2101年度予算", slug="mahoro-city-2101", government=gov, classification_system=cs
         )
@@ -196,6 +199,29 @@ class WdmmgTestCase(BudgetMapperTestUserAPITestCase):
                                     "name": cl100.name,
                                     "code": cl100.code,
                                     "amount": 131.0,
+                                    "children": None,
+                                }
+                            ],
+                        }
+                    ],
+                },
+                {
+                    "id": cl2.id,
+                    "name": cl2.name,
+                    "code": cl2.code,
+                    "amount": 0,
+                    "children": [
+                        {
+                            "id": cl20.id,
+                            "name": cl20.name,
+                            "code": cl20.code,
+                            "amount": 0,
+                            "children": [
+                                {
+                                    "id": cl200.id,
+                                    "name": cl200.name,
+                                    "code": cl200.code,
+                                    "amount": 0,
                                     "children": None,
                                 }
                             ],

